ndk/src/surface_texture.rs

```diff
@@ -0,0 +1,126 @@
+//! Bindings for [`ffi::ASurfaceTexture`]
+#![cfg(feature = "api-level-28")]
+
+use super::posix::PosixError;
+use crate::native_window::NativeWindow;
+use jni_sys::{jobject, JNIEnv};
+use std::ptr::NonNull;
+
+#[derive(Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct SurfaceTexture {
+    ptr: NonNull<ffi::ASurfaceTexture>,
+}
+
+unsafe impl Send for SurfaceTexture {}
+
+impl Drop for SurfaceTexture {
+    fn drop(&mut self) {
+        unsafe { ffi::ASurfaceTexture_release(self.ptr.as_ptr()) }
+    }
+}
+
+impl SurfaceTexture {
+    /// Assumes ownership of `ptr`
+    ///
+    /// # Safety
+    /// `ptr` must be a valid pointer to an Android [`ffi::ASurfaceTexture`].
+    pub unsafe fn from_ptr(ptr: NonNull<ffi::ASurfaceTexture>) -> Self {
+        Self { ptr }
+    }
+
+    /// # Safety
+    ///
+    /// This function should be called with a healthy JVM pointer and with a non-null surface texture,
+    /// which must be kept alive on the Java/Kotlin side.
+    /// Get a reference to the native ASurfaceTexture from the corresponding java object.
+    //
+    /// The caller must keep a reference to the Java `SurfaceTexture` during the lifetime of the returned [`SurfaceTexture`].
+    /// Failing to do so could result in the [`SurfaceTexture`] to stop functioning properly once the Java object gets finalized.
+    /// However, this will not result in program termination.
+    pub unsafe fn from_surface_texture(env: *mut JNIEnv, surface_texture: jobject) -> Option<Self> {
+        let a_surface_texture_ptr = ffi::ASurfaceTexture_fromSurfaceTexture(env, surface_texture);
+        let s = NonNull::new(a_surface_texture_ptr)?;
+        Some(SurfaceTexture::from_ptr(s))
+    }
+
+    /// Returns a pointer to the native [`ASurfaceTexture`](ffi::ASurfaceTexture).
+    pub fn ptr(&self) -> NonNull<ffi::ASurfaceTexture> {
+        self.ptr
+    }
+
+    /// Returns a reference to a [`NativeWindow`] (i.e. the Producer) for this [`SurfaceTexture`].
+    ///
+    /// This is equivalent to Java's:
+    /// ```java
+    /// Surface sur = new Surface(surfaceTexture);
+    /// ```
+    pub fn acquire_native_window(&mut self) -> Option<NativeWindow> {
```

Comment:
I don't think this needs to be `mut`, it seems I can safely crate and use multiple `NativeWindow`s (`Surface`, producers) on a single `SurfaceTexture` and the result doesn't hold a (mutable) borrow on `self` anyway.

Otherwise `update_tex_image` should probably be `mut` as well based on the name, but it should all be internally synchronize so there should be very little reason to force mutable (unique, really) access here.